contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,555 @@
+     *
+     * @param _transformComponents      array of components to transform
+     * @param _transformData            array of transform data 
+     */
+    function batchTransform(
+        address[] memory _transformComponents,
+        bytes[] memory _transformData
+    )
+        external
+        onlyAllowedCaller(msg.sender)
+    {
+        require(_transformComponents.length == _transformData.length, "length mismatch");
+
+        _absorbAirdrops(_transformComponents);
+
+        for (uint256 i = 0; i < _transformComponents.length; i++) {
+            _transform(_transformComponents[i], _transformData[i], false);
+        }
+    }
+
+    /**
+     * ONLY ALLOWED CALLER: Transforms all remaining units of the underlying component into the transfrorm component. This function should
+     * only be called if it is the final transformation for the underlying, and the raw underlying is not an intended component of the final
+     * Set composition. The transformData parameter can be fetched by the rebalance bots by calling getTransformData on the relevent
+     * TransformHelper.
+     *
+     * @param _transformComponent       component to transform
+     * @param _transformData            transform data
+     */
+    function transformRemaining(
+        address _transformComponent,
+        bytes memory _transformData
+    )
+        external
+        onlyAllowedCaller(msg.sender)
+    {
+        address underlying = transformComponentInfo[_transformComponent].underlyingComponent;
+        uint256 underlyingUnitsRaw = targetUnitsUnderlying[underlying];
+        require(underlyingUnitsRaw == 0, "raw underlying in target set composition");
+
+        _transform(_transformComponent, _transformData, true);
+    }
+
+    /* ======== Internal Functions ======== */
+
+    /**
+     * Untransforms a component. To prevent losses from untransforms that may incur slippage, a maximum
+     * untransform size and minimum untransform delay are enforced. 
+     * 
+     * To calculate the number of units to untransform use:
+     * max(currentUnits - targetUnitsUnderlying * exchangeRate, 0)
+     *
+     * This calculation ensures that if the target units is greater than the current units then no tokens will
+     * be untransformed. If this calculation yields a resulting amount to untransform that is larger than the
+     * maximum untransform size, the untransform size will be replaced with this max value. The rest of the
+     * transform component can be untransformed in later transactions.
+     */
+    function _untransform(address _transformComponent, bytes memory _untransformData) internal {
+        require(stage == RebalanceStage.UNTRANSFORM, "must be in untransform stage");
+
+        TransformInfo memory transformInfo = transformComponentInfo[_transformComponent];
+
+        require(transformInfo.minTransformDelay + lastTransform[_transformComponent] < block.timestamp, "delay not elapsed");
```

Comment:
All these requires can be grouped together and moved to an internal function, `validateUntransform`